apps/passport-server/src/services/devconnectPretixSyncService.ts

```diff
@@ -0,0 +1,498 @@
+  insertPretixItemsInfo,
+  updatePretixItemsInfo,
+} from "../database/queries/pretixItemInfo";
+import { ApplicationContext } from "../types";
+import {
+  getEmailAndItemKey,
+  pretixTicketsDifferent,
+  ticketsToMapByEmailAndItem,
+} from "../util/devconnectTicket";
+import { logger } from "../util/logger";
+import { RollbarService } from "./rollbarService";
+import { SemaphoreService } from "./semaphoreService";
+import { traced } from "./telemetryService";
+
+const SERVICE_NAME_FOR_TRACING = "Devconnect Pretix";
+
+/**
+ * Responsible for syncing users from Pretix into an internal representation.
+ */
+export class DevconnectPretixSyncService {
+  private pretixAPI: IDevconnectPretixAPI;
+  private rollbarService: RollbarService | null;
+  private semaphoreService: SemaphoreService;
+  private context: ApplicationContext;
+  private timeout: NodeJS.Timeout | undefined;
+  private _hasCompletedSyncSinceStarting: boolean;
+
+  public get hasCompletedSyncSinceStarting(): boolean {
+    return this._hasCompletedSyncSinceStarting;
+  }
+
+  public constructor(
+    context: ApplicationContext,
+    pretixAPI: IDevconnectPretixAPI,
+    rollbarService: RollbarService | null,
+    semaphoreService: SemaphoreService
+  ) {
+    this.context = context;
+    this.rollbarService = rollbarService;
+    this.semaphoreService = semaphoreService;
+    this.pretixAPI = pretixAPI;
+    this._hasCompletedSyncSinceStarting = false;
+  }
+
+  public replaceApi(newAPI: IDevconnectPretixAPI): void {
+    const wasRunning = !!this.timeout;
+
+    if (wasRunning) {
+      this.stop();
+    }
+
+    this.pretixAPI = newAPI;
+    this._hasCompletedSyncSinceStarting = false;
+
+    if (wasRunning) {
+      this.startSyncLoop();
+    }
+  }
+
+  public startSyncLoop(): void {
+    const trySync = async (): Promise<void> => {
+      await this.trySync();
+      this.timeout = setTimeout(() => trySync(), 1000 * 60);
+    };
+
+    trySync();
+  }
+
+  public async trySync(): Promise<void> {
+    try {
+      await this.sync();
+      await this.semaphoreService.reload();
+      this._hasCompletedSyncSinceStarting = true;
+    } catch (e) {
+      this.rollbarService?.reportError(e);
+      logger(e);
+    }
+  }
+
+  public stop(): void {
+    if (this.timeout) {
+      clearTimeout(this.timeout);
+    }
+  }
+
+  /**
+   * Synchronize Pretix state with PCDPassport state.
+   */
+  private async sync(): Promise<void> {
+    return traced(SERVICE_NAME_FOR_TRACING, "sync", async () => {
+      const syncStart = Date.now();
+      logger("[DEVCONNECT PRETIX] Sync start");
+
+      const { dbPool } = this.context;
+
+      const promises = [];
+      for (const organizer of this.pretixAPI.config.organizers) {
+        for (const event of organizer.events) {
+          promises.push(
+            this.syncAllPretixForOrganizerAndEvent(dbPool, organizer, event)
+          );
+        }
+      }
+      try {
+        await Promise.all(promises);
```

Comment:
we could be a bit more explicit about what happens when any of these syncs failed

Reply:
added a comment and a bit more info to the logs